text/0000-stdlib-term-mouse-input-plus.md

```diff
@@ -0,0 +1,196 @@
+- Feature Name: stdlib-term-mouse-input-plus
+- Start Date: 2022-06-20
+- RFC PR: 
+- Pony Issue: 
+
+## Summary
+
+Enhance ANSI terminal support in package `term` with mouse input handling, additional ANSI escape codes to support alt buffer and cursor switching, and capture of SIGINT and SIGTSTP to handle `^C` and `^Z` as input.
+
+**Table of contents**
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Detailed design](#detailed-design)
+  - [New `ANSINotify` methods](#new-ansinotify-methods)
+  - [New `ANSI` escape codes](#new-ansi-escape-codes)
+  - [Enhance `ANSITerm`](#enhance-ansiterm)
+    - [Mouse input handling](#mouse-input-handling)
+    - [Ctrl-C and Ctrl-Z input handling](#ctrl-c-and-ctrl-z-input-handling)
+  - [ANSI terminal options](#ansi-terminal-options)
+- [How We Teach This](#how-we-teach-this)
+- [How We Test This](#how-we-test-this)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Reference Implementation](#reference-implementation)
+- [Questions](#questions)
+- [References](#references)
+
+## Motivation
+
+> Why are we doing this? 
+
+Improve ANSI terminal application capabilities when implementing text-based UIs.
+
+> What use cases does it support? 
+
+Mouse input handling, alt buffer switching, cursor hiding and capturing missing control keys.
+
+> What is the expected outcome?
+
+Enable richer text UI applications with both key and mouse input and improved user experience.
+
+## Detailed design
+
+> This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the language to understand, and for somebody familiar with the compiler to implement. This should get into specifics and corner-cases, and include examples of how the feature is used.
+
+### New `ANSINotify` methods
+
+`fun ref mouse_down(button: U8, x: U32, y: U32)`
+* This method is called whenever a mouse button is pressed
+* `button` is either 0 (left), 1 (middle), 2 (right)
+* `x` and `y` are the character coordinates below the mouse cursor
+
+`fun ref mouse_up(button: U8, x: U32, y: U32) => None` 
+* This method is called whenever a mouse button is released
+* `button` is either 0 (left), 1 (middle), 2 (right)
+* `x` and `y` are the character coordinates below the mouse cursor
+
+`fun ref mouse_move(x: U32, y: U32) => None`
+* This method is called whenever the mouse is moved while no button is pressed
+* `x` and `y` are the character coordinates below the mouse cursor
+
+`fun ref mouse_drag(button: U8, x: U32, y: U32) => None`
+* This method is called whenever the mouse is moved while a button is pressed
+* `button` is either 0 (left), 1 (middle), 2 (right)
+* `x` and `y` are the character coordinates below the mouse cursor
+
+`fun ref mouse_wheel(direction: U8, x: U32, y: U32) => None`
+* This method is called whenever the mouse wheel is rolled in either direction. 
+* `direction` is either 0 or 1
```

Comment:
@rhagenson, ideally direction is UP or DOWN, but given that we now have two kinds of up/down because of "natural" vs what we had before swiping happened, up or down is relative. Unlike the buttons, in this case there are only two values so we could define `Up` and `Down` ... though those might be too common to put in the global namespace. Do you have a recommendation for the names?

Reply:
What do you mean by global namespace?

Reply:
I just meant that after a `use "term"` statement, any primitives defined within would become accessible to the rest of the code in that file, and thus a `primitive Up fun apply() : U8 => 1` would put the name `Up` into the namespace. (It's how I think about it, though in hindsight "global" is likely not the right word since I assume that's reserved for the built-in objects.)

Reply:
Don't take that into account when naming things. You can always alias names when importing so using a common name shouldn't be something you consider.